### ISSUE TITLE
Fare Hacking on BART  Stephen Brennan

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4393]
Fare Hacking on BART &bull; Stephen Brennan<br>
<br>
http://bit.ly/2IlmYPz<br>
<br>
Stephen Brennan<br>
<br>
<div>
<div>
<h1>Fare Hacking on BART</h1>
<em>Stephen Brennan &bull; 23 July 2016</em>
<p>Imagine that you&rsquo;re taking a long train ride on the BART. Maybe from Millbrae to North Concord. Chances are, at the very same moment, somebody else is going the other direction. For example, maybe from Pittsburgh/Bay Point to San Bruno. In case you haven&rsquo;t memorized BART&rsquo;s stops, here&rsquo;s a useful map illustrating these rides:</p>
<p><img src="https://brennan.io/images/bart-labeled.png" alt="bart-labeled"></p>
<p>You&rsquo;re the cyan rider (starting at the chevron and ending at the octagon), and the other person is in magenta.</p>
<p>Now, if you both bought correct tickets, your trips would cost a little more than $7 each. That&rsquo;s not unreasonable, but there&rsquo;s a way you could pay less, so long as you know the Magenta rider.</p>
<ol>
<li>Instead of buying a ticket for Millbrae to North Concord, you just buy a ticket for Millbrae to San Bruno.</li>
<li>Magenta buys a ticket for Pittsburgh/Bay Point to North Concord.</li>
<li>Halfway through, you both get off your trains, meet on the platform, and swap your tickets.</li>
<li>Then you continue on the next train to your destinations.</li>
</ol>
<p>This ends up saving both of you around $5, but in the process you lose 20 minutes standing around on a platform in Oakland! Probably not worth it. But, what if there were a way to do this electronically so that you never had to get off the train? After all, Clipper Cards (the electronic ticket of choice on the BART system) use NFC, which is a technology built into some smartphones. Although Clipper Cards currently aren&rsquo;t cloneable or spoofable, if they were, an app could theoretically use the principle from above to swap tickets around, reducing everyone&rsquo;s fares!</p>
<p>I&rsquo;ve had this idea bouncing around in my head for a couple weeks. Thursday and Friday were Hackathon days at Yelp, which gave me the perfect opportunity to take a shot at making this app. I successfully implemented what I think is a pretty neat algorithm and API, and then ran it on some real BART data. My results indicate that an app like this could save its users between 20 and 40% on BART fares, depending on time of day. Read on for all the gory details&mdash;I think it&rsquo;s pretty fascinating!</p>
<p>One final note before I dive in: this algorithm is essentially large scale ticket swapping, which is obviously illegal and unethical. I don&rsquo;t believe that it would be ethical to use this in the real world. I value safe and smooth travel more than I value the 20-40% of my fare I could save by stealing from BART. With that said, I still think this is a really cool problem, and since Clipper cards are not cloneable, this work can&rsquo;t be used to facilitate this large-scale theft.</p>
<h2 id="the-algorithm">The algorithm</h2>
<p>In order to write an algorithm, we should always step back and formulate the problem as clearly as we can. In this case, the input to our problem is a bunch of people who would like to go from point A to point B. Our job is to &ldquo;purchase&rdquo; tickets that we can assign to people on entry, and then swap to other people on exit, so that we minimize the total cost of these tickets.</p>
<p>Let&rsquo;s go back to the toy example I presented in the beginning. Why does it work? What is the fundamental property we are taking advantage of? It&rsquo;s actually pretty simple: BART would like to enforce that everyone who travels from point A to point B buys a ticket for that source and destination. However, with just turnstiles at stations, they can&rsquo;t enforce that. The best they can enforce is that each person entering a station has a valid ticket for a trip starting there, and that each person exiting a station has a valid ticket for a trip ending there.</p>
<p>This gives us a set of constraints. Our algorithm can purchase whatever tickets it wants, so long as we end up with the same number of people entering and leaving each station. Since we&rsquo;re minimizing the total cost of the tickets, we have a classic optimization problem. In particular, this problem can be expressed as an <strong>Integer Linear Program</strong>.</p>
<p><a href="https://en.wikipedia.org/wiki/Linear_programming">Linear programs</a> are math problems where you are trying to choose values for a vector (i.e. a list) of variables <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png">, such that you minimize a cost function. Typically, each variable <img src="https://brennan.io/res/latex/latex-88bb695553de7c8ef3800e6de080d09f.png"> in <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png"> has an associated cost <img src="https://brennan.io/res/latex/latex-7483370cbe7d293bf85433e89f27498d.png">, and so the cost function is just the sum of the <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png">&rsquo;s times their costs. But, you have to satisfy some constraints, which are usually expressed as equations, like this:</p>
<p><img src="https://brennan.io/res/latex/latex-8594b21182e0502cebbcbdd702b74f81.png"></p>
<p>We can have lots of constraints we need to satisfy, so we typically number the constraints from one to <img src="https://brennan.io/res/latex/latex-31db899bf8a4e7072a271cef91a878df.png">. We can compactly write all the constraints using a matrix form like this:</p>
<p><img src="https://brennan.io/res/latex/latex-9992ac8cbd2235e96add81b0c00e771e.png"></p>
<p>Here, <img src="https://brennan.io/res/latex/latex-69d94549c96095ed19144dd989e34a99.png"> is an <img src="https://brennan.io/res/latex/latex-31db899bf8a4e7072a271cef91a878df.png"> row (one for each constraint) by <img src="https://brennan.io/res/latex/latex-fc2f7800ad464513027725921007c268.png"> column (one for each variable) matrix containing the coefficients <img src="https://brennan.io/res/latex/latex-51c163119c2b33dc900381a8223e8253.png"> from all the constraints. Usually we also have the constraint that <img src="https://brennan.io/res/latex/latex-f0712e3a9cd9bff88afc646f6c4e949e.png">, and for integer linear programs, we also need to make sure that our solutions for <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png"> are integers. Since linear programs are pretty common problems to solve, there are plenty of existing solvers that can solve them <em>reasonably</em> quickly. If you can write a problem as a linear program and come up with <img src="https://brennan.io/res/latex/latex-69d94549c96095ed19144dd989e34a99.png">, <img src="https://brennan.io/res/latex/latex-7c90cbff19a699fdec5910748c481aeb.png">, and the costs <img src="https://brennan.io/res/latex/latex-437329dd831483c024186dc19f207f3c.png">, you can use these linear programming libraries to solve your problem for you. So let&rsquo;s formulate this as an integer linear program!</p>
<p>In this particular problem, we have a variable <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png"> for every single pair of stations you could start at and end at. <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png"> will represent how many tickets we buy for that start/end pair. There are 45 BART stations, which means that we have <img src="https://brennan.io/res/latex/latex-6f225f8cb05a2a2ea4e1315108ecdbae.png"> different variables. The cost for each variable is provided in a fare schedule, available from BART&rsquo;s <a href="https://www.bart.gov/sites/default/files/docs/2016%20Fare%20Chart.pdf">website</a>. The one value that is missing from this schedule is the cost of a trip that starts and ends at the same station. You&rsquo;d think that this would be free, but people would take advantage of that. So BART charges $5.75 for this and markets it as a chance to &ldquo;explore&rdquo;<sup id="fnref:explore"></sup> their train system!</p>
<p>I already outlined the constraints in words, and now we can actually express them mathematically. The first constraint was that the number of tickets starting at any station has to be equal to the number of travelers starting at that station. So, the constraint is:</p>
<p><img src="https://brennan.io/res/latex/latex-6819d870dbf1ecf0bacba618c738850f.png"></p>
<p>We have one of these constraints for each station. We&rsquo;ll call them the <em>source</em> constraints. The second type of constraint was that the number of tickets ending at any station has to be equal to the number of travelers who want to go to that station:</p>
<p><img src="https://brennan.io/res/latex/latex-81fa66150170fe4d7e8d4755e37ba739.png"></p>
<p>And these are our <em>destination</em> constraints. Together, the source and destination constraints can be represented in using the matrix form <img src="https://brennan.io/res/latex/latex-e1e2079dab89d32027d826c84500056a.png">. The matrix is pretty large, with 90 rows and 2025 columns! Regardless of how big it is, this should mean that we&rsquo;re done. We can use a linear programming library to solve the problem.</p>
<h3 id="integer-linear-programs-are-hard">Integer linear programs are hard</h3>
<p>Actually, we may not yet be done. You see, while linear programs are solvable on average in a reasonable (a.k.a. polynomial) amount of time, integer linear programs are not. They are NP-hard, which means that you can&rsquo;t expect to find a solution in a reasonable amount of time.</p>
<p>However, sometimes a particular instance of an integer linear program is not that hard. For some problems, if you just &ldquo;forget&rdquo; the constraint that your solutions should be integers, and solve the problem as a normal linear program, you&rsquo;ll always get integer solutions anyway.</p>
<p>This happens whenever your constraint matrix <img src="https://brennan.io/res/latex/latex-69d94549c96095ed19144dd989e34a99.png"> is <a href="https://en.wikipedia.org/wiki/Unimodular_matrix">totally unimodular</a>. I won&rsquo;t explain what that is, but it turns out that our matrix is unimodular!<sup id="fnref:fn-unimodular"></sup> So this means that we <em>can</em> just use a linear programming library to solve it. In fact, here is the code I used to do it:</p>
<div>
<div>
<pre>
<code><span>b</span> <span>=</span> <span>np</span><span>.</span><span>hstack</span><span>([</span><span>src_sum</span><span>(</span><span>self</span><span>.</span><span>traveler_matrix</span><span>),</span> <span>dst_sum</span><span>(</span><span>self</span><span>.</span><span>traveler_matrix</span><span>)])</span><span>.</span><span>astype</span><span>(</span><span>np</span><span>.</span><span>float</span><span>)</span> <span>A_src_const</span> <span>=</span> <span>np</span><span>.</span><span>repeat</span><span>(</span><span>np</span><span>.</span><span>identity</span><span>(</span><span>self</span><span>.</span><span>num_stations</span><span>),</span> <span>self</span><span>.</span><span>num_stations</span><span>,</span> <span>axis</span><span>=</span><span>1</span><span>)</span> <span>A_dst_const</span> <span>=</span> <span>np</span><span>.</span><span>hstack</span><span>(</span> <span>[</span><span>np</span><span>.</span><span>identity</span><span>(</span><span>self</span><span>.</span><span>num_stations</span><span>)</span> <span>for</span> <span>_</span> <span>in</span> <span>range</span><span>(</span><span>self</span><span>.</span><span>num_stations</span><span>)])</span> <span>A</span> <span>=</span> <span>np</span><span>.</span><span>vstack</span><span>([</span><span>A_src_const</span><span>,</span> <span>A_dst_const</span><span>])</span> <span>c</span> <span>=</span> <span>self</span><span>.</span><span>fare_matrix</span><span>.</span><span>reshape</span><span>(</span><span>self</span><span>.</span><span>num_stations</span> <span>**</span> <span>2</span><span>)</span> <span>self</span><span>.</span><span>res</span> <span>=</span> <span>scipy</span><span>.</span><span>optimize</span><span>.</span><span>linprog</span><span>(</span><span>c</span><span>,</span> <span>A_eq</span><span>=</span><span>A</span><span>,</span> <span>b_eq</span><span>=</span><span>b</span><span>)</span> </code>
</pre>
</div>
</div>
<p>There is some stuff missing&mdash;you can see the code in its original context <a href="https://github.com/brenns10/bart/blob/c148563af68b68237ad4556e0111b6bc3052ed59/bart/solver.py#L82">here</a>. The point of this sample is to demonstrate that formulating your problem as a linear program can make for a very nice, compact, and clean solution! However, the limitation is that your solution is only as good as the underlying linear programming library!</p>
<h3 id="regular-linear-programming-is-hard-too">Regular linear programming is hard too</h3>
<p><del>It turns out that not every linear programming library is created equal. Sometimes, they don&rsquo;t work right. Since my project was in Python, I used NumPy and SciPy, the industry and academic standard toolchain for math in Python. SciPy&rsquo;s implementation of the <a href="https://en.wikipedia.org/wiki/Simplex_algorithm">simplex</a> algorithm, like most implementations, has two phases. In the first, it searches for any solution that can satisfy the constraints (this is called a basic feasible solution). In the second, it takes this solution and modifies it until it finds the optimal solution. However, occasionally SciPy wasn&rsquo;t able to find a basic feasible solution in the first step, even though in this case there is always a very simple one: just buy a ticket for each person&rsquo;s ride directly. It&rsquo;s not optimal, but it satisfies the source and destination constraints. Unfortunately, I wasn&rsquo;t really able to get around this problem, so it was right back to square one for me.</del></p>
<p><strong>Edit (December 2016):</strong> My solution using SciPy&rsquo;s linear programming solver did not work, because I sometimes gave it linearly dependent constraints. I wasn&rsquo;t aware that was the problem. You can read more about this issue and how I&rsquo;ve resolved it <a href="https://brennan.io/2016/12/11/bart-revisited/">here</a>.</p>
<p>But this was a hackathon, and there&rsquo;s no giving up at a hackathon! If one thing doesn&rsquo;t work, you just find a new way to do it. And you don&rsquo;t waste too much time going down rabbit holes trying to fix it. So instead I decided to come up with a custom algorithm from scratch for this problem.</p>
<h2 id="the-other-algorithm">The other algorithm</h2>
<p>So even though this problem can be expressed as a linear program doesn&rsquo;t mean that you have to solve it with one. In fact, simply knowing that a problem can be written as a linear program can give you a lot of insight into how you should write a custom algorithm for solving it. For instance, a problem that can be expressed as a linear program is convex, which is very convenient. This means that we can use a simple algorithm to solve it.</p>
<p>To understand why, think of a very simple &ldquo;optimization&rdquo; problem: finding the highest point in an area of land. If the area of land you&rsquo;re on has a single hill, it&rsquo;s probably convex, meaning that there are no valleys or other indented regions. This means you can follow the straightforward strategy of always following the steepest slope. No matter what, this approach is <em>guaranteed</em> to get you to the top of the hill.</p>
<p><img src="https://brennan.io/images/bart-one-hill.jpg" alt="bart-one-hill"></p>
<p><em>Above: Windows XP&rsquo;s desktop background, representing a convex optimization problem. There is only one peak. <a href="http://www.extremetech.com/wp-content/uploads/2014/04/bliss-windows-xp-original.jpg">Source</a></em></p>
<p>On the other hand, if the area you&rsquo;re on has multiple hills, it&rsquo;s definitely not convex: there&rsquo;s a valley in between! If you follow the same strategy as above, you might end up on top of the highest hill, but you also might end up on a different peak: a <em>local optimum</em>.</p>
<p><img src="https://brennan.io/images/bart-multiple-hills.jpg" alt="bart-multiple-hills"></p>
<p><em>Above: An non-convex area with many &ldquo;local&rdquo; peaks. <a href="http://static5.techinsider.io/image/56059ce8bd86ef21008bd25e-1190-625/north-americas-tallest-mountain-is-shorter-than-we-thought.jpg">Source</a></em></p>
<p>These examples may seem silly or contrived, but in fact many optimization problems boil down to something as simple as &ldquo;climbing a hill&rdquo;, but instead of changing your X and Y coordinates, you can change many more variables: approximately 2025 in the case of the problem of BART fare minimization. But since this problem is convex, it&rsquo;s rather simple: like finding the top of a single hill, but in 2025 dimensional space!</p>
<p>So how do we find our way up this 2025 dimensional hill? To answer that question, let&rsquo;s represent our problem a bit differently: with a matrix. We&rsquo;ll label both the rows and the columns with stations: the rows are the starting stations, and the columns are ending stations. A number <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png"> in row <img src="https://brennan.io/res/latex/latex-a796859cf0d2898de27a0ef592a69b6d.png">, column <img src="https://brennan.io/res/latex/latex-f30a01ea9ef93c2932cbe9e00f4add20.png"> means that <img src="https://brennan.io/res/latex/latex-7783896d2879a1b45ee97c4ffe0a4531.png"> people are riding from station <img src="https://brennan.io/res/latex/latex-a796859cf0d2898de27a0ef592a69b6d.png"> to station <img src="https://brennan.io/res/latex/latex-f30a01ea9ef93c2932cbe9e00f4add20.png">. We&rsquo;ll call this our traveler matrix, <img src="https://brennan.io/res/latex/latex-33aae57de23c92db9c5754d5a2caf87f.png">. Here&rsquo;s a small example with four stations:</p>
<p><img src="https://brennan.io/res/latex/latex-3e7b49c5affceb5407dd1c66b15bd5b5.png"></p>
<p>In this example, one traveler wants to go from station 1 to 3, and another wants to go from station 4 to 2. If you imagine the stations in a line, it&rsquo;s a bit easier to visualize.</p>
<p><img src="https://brennan.io/images/bart-number-line.jpg" alt="bart-number-line"></p>
<p>If you imagine that the fare is $1 for every &ldquo;link&rdquo; in this simple train route, then the fare matrix <img src="https://brennan.io/res/latex/latex-0c38c5c84e82743e30c2c48ff79b481d.png"> would look something like this:</p>
<p><img src="https://brennan.io/res/latex/latex-d4012192dc56301fc224ad3e5cbc7ff2.png"></p>
<p>You could get the total fare for everyone in this little system by taking each element in the &ldquo;traveler matrix&rdquo; and multiplying it with its corresponding element in the &ldquo;fare matrix&rdquo;, then adding them all up<sup id="fnref:fn-frobenius"></sup>. For the example travelers, the total fare is $4.</p>
<p>According to our constraints, we are free to do anything we&rsquo;d like to the traveler matrix (which we can refer to now as a ticket matrix, since it will represent the tickets our algorithm buys), so long as we maintain the source and destination constraints. In this matrix formulation of the problem, this simply means that we can&rsquo;t do anything to the matrix that changes its row or column sums.</p>
<p>There&rsquo;s a very convenient way to change a matrix so that you don&rsquo;t modify the row and column sums: simply find two entries in different rows and columns that are greater than zero. They form two corners of a rectangle in the matrix. Subtract however many tickets you want from both those entries, and add them to the other two corners of the rectangle. Let&rsquo;s do this with the previous example <img src="https://brennan.io/res/latex/latex-33aae57de23c92db9c5754d5a2caf87f.png">:</p>
<p><img src="https://brennan.io/res/latex/latex-a24d6a07c2c76ab032a9835fce62eb39.png"></p>
<p>If we look back at the fare matrix, we can see that this transformation actually lowered the total fare of the system from 4 to 2! This is because we went from having tickets <img src="https://brennan.io/res/latex/latex-a6608f7f432c7a7037da96e06be054c3.png"> and <img src="https://brennan.io/res/latex/latex-23d16f5368ab4923494a825997479251.png">, which both cost $2 to simply having tickets <img src="https://brennan.io/res/latex/latex-fbb163b0c4aa77864707d2cee9076e46.png"> and <img src="https://brennan.io/res/latex/latex-1e9e84821d76c0388ab4448e683855f9.png">, which both cost $1.</p>
<p>And in fact, that&rsquo;s pretty much the whole algorithm! All my custom algorithm does is look in the matrix for pairs where it can perform this swap, and do it if it will lower the total fare. Since lowering the fare is the equivalent of going &ldquo;uphill&rdquo; in our 2025 dimensional problem, we know that as long as we keep going uphill, we&rsquo;ll eventually reach the peak. Thank you, convexity!</p>
<p>This algorithm, as I implemented it, is not very smart or efficient, but it&rsquo;s definitely &ldquo;hackathon complete&rdquo;. It does the job!</p>
<h2 id="the-implementation">The implementation</h2>
<p>After I got this algorithm settled, the next order of business was implementing it in a realistic way: as if it were the basis of a real app that would be deployed to real users. So I made a REST API, which clients (apps) can use. The most important endpoint is <code>/travel</code>. Whenever a client sends a request to <code>/travel</code>, it provides a source, destination, and a piece of identifying info, like a name or a clipper card number. The server adds that traveler to the current &ldquo;batch&rdquo; of travelers waiting to have their tickets swapped. The server returns a unique ID to each traveler which they can look up later to get their exit ticket. Then, at some point an admin requests the <code>/calculate</code> endpoint, which instructs the server to perform the ticket swapping algorithm on the current batch of travelers. The server calculates it all and returns the original and optimal fare cost. After that, travelers can look up their ID on the <code>/result/&lt;id&gt;</code> endpoint to get their exit ticket.</p>
<p>The only client I created was an &ldquo;admin&rdquo; website. It has two panels. The first is a manual entry form, which allows you to manually add travelers and then do the fare swapping algorithm. Here you can see it working on the example from the beginning of this article:</p>
<p><img src="https://brennan.io/images/bart-admin.png" alt="bart-admin"></p>
<p>The second is a simulation interface. It allows you to upload a day&rsquo;s worth of BART origin-destination data in CSV format (available <a href="http://www.bart.gov/about/reports/ridership">here</a>). It will then send this all to the server, running the algorithm once for each hour of data. Essentially it can simulate a whole day of real BART traffic going into the app! As the simulation progresses, the site graphs a few performance measures to see how well the algorithm reduces fares. The results are pretty fascinating!</p>
<h2 id="the-results">The results</h2>
<p>I did two different simulations: one on a typical weekday with rush hour commute patterns (Thursday, August 6, 2015) and one on a weekend (Saturday, August 8, 2015). Below is the &ldquo;savings&rdquo; graph my simulation produced on the weekday. It plots the amount of money riders would save each hour if they all were using my fare swapping algorithm.</p>
<p><img src="https://brennan.io/images/bart-weekday-savings.png" alt="bart-weekday-savings"></p>
<p>You can very clearly see two peaks during rush hour traffic. Even more exciting are the raw numbers: the graph peaks at $50,000 saved for a single hour during afternoon rush hour. During most operating hours the algorithm saves at least $20,000 per hour. Note that these savings assume that riders are paying the fare price listed on the fare schedule, passes are not taken into account here.</p>
<p>Perhaps more important than knowing the amount of money saved is knowing what fraction of their original fares riders would pay with the ticket swapping method:</p>
<p><img src="https://brennan.io/images/bart-weekday-fraction.png" alt="bart-weekday-fraction"></p>
<p>It appears that the savings fluctuate between 20% and 40%. If you look carefully, you&rsquo;ll notice that the regions of the graph with 20% savings correspond to the rush hour traffic peaks from above. Savings are probably less during rush hour because most of the traffic is directed between work areas and residential areas. There are fewer swaps the algorithm can make in this situation.</p>
<p>This hypothesis is supported by looking at the same plot from the weekend simulation:</p>
<p><img src="https://brennan.io/images/bart-weekend-fraction.png" alt="bart-weekend-fraction"></p>
<p>The savings are between 30% and 40% on this plot, likely because the traffic doesn&rsquo;t have the directed flow that rush hour has.</p>
<h2 id="conclusion">Conclusion</h2>
<p>Over the course of Thursday and Friday, I came up with an algorithm for swapping BART riders&rsquo; tickets to minimize their fares. I implemented it, wrapped it in a rough &ldquo;API&rdquo;, and evaluated its performance on real BART ridership data. The results? A hefty 20-40% discount on your train rides, and some neat visualizations of BART traffic patterns to boot.</p>
<p>Would it work in the real world? Maybe, if you assume that a mechanism for &ldquo;cloning&rdquo; Clipper Cards could become available some day. Disregarding that major barrier, there are still some problems that would have to be solved in a real world application of this algorithm:</p>
<ul>
<li>
<strong>Adoption:</strong> The simulation results are so rosy because they assume everyone participates in the algorithm. Like social networks, the app might not be very good without a lot of users.</li>
<li>
<strong>Scheduling:</strong> My simulation ran the algorithm on a full hour of riders. In reality, you&rsquo;d need to run it on smaller time increments (no larger than 10 minutes) so that riders wouldn&rsquo;t be waiting on the app just to leave the station. I have plenty of interesting ideas for how to address this problem, but not enough hackathon time to try them.</li>
<li>
<strong>Performance:</strong> My algorithm is rather slow right now (a day&rsquo;s simulation took around 40 minutes, although that does include some overhead from API requests). There are plenty of low hanging fruit for speeding it up, but I didn&rsquo;t have time during the hackathon.</li>
</ul>
<p>In any case, since I have no idea of launching a &ldquo;stealing from BART&rdquo; app any time soon, these &ldquo;future work&rdquo; problems will have to go unsolved.</p>
<p>This was a really enjoyable problem to solve. I learned a lot from it, and I got to apply some concepts that you normally just learn about in school. Hopefully people will enjoy reading about it as much as I enjoyed doing it!</p>
<p>If you want to learn more, check out the code on <a href="https://github.com/brenns10/bart">GitHub</a>.</p>
<p><strong>Edit (December 2016):</strong> An update to this post can be found <a href="https://brennan.io/2016/12/11/bart-revisited/">here</a>.</p>
<hr>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Stephen Brennan's Blog https://brennan.io<br>
<br>
June 17, 2019 at 11:46AM
